lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,509 @@
+            SetErrMsg(message.c_str());
+            return ERROR;
+        }
+        return FOUND;
+    }
+    return NOT_FOUND;
+}
+
+// Template specialization for reading data of type std::array<size_t, 3> or std::array<float, 3>
+template<typename T> int BOVCollection::_findToken(const std::string &token, std::string &line, std::array<T, 3> &value, bool verbose)
+{
+    // Skip comments
+    if (line[0] == '#') { return NOT_FOUND; }
+
+    size_t pos = line.find(token);
+    if (pos != std::string::npos) {    // We found the token
+        T lineValue;
+        _findTokenValue(line);
+        std::stringstream lineStream(line);
+
+        for (int i = 0; i < value.size(); i++) {
+            lineStream >> lineValue;
+            value[i] = lineValue;
+        }
+
+        if (lineStream.bad()) {
+            // value.clear();
+            std::string message = "Invalid value for " + token + " in BOV header file.";
+            SetErrMsg(message.c_str());
+            return ERROR;
+        }
+
+        if (verbose) {
+            std::cout << std::setw(20) << token << " ";
+            for (int i = 0; i < value.size(); i++) std::cout << value[i] << " ";
+            std::cout << std::endl;
+        }
+        return FOUND;
+    }
+    return NOT_FOUND;
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+int BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(std::string varname, size_t ts, const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    float       time = _times[ts];
+    std::string dataFile = _dataFileMap[varname][time];
+
+    FILE *fp = fopen(dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %d", fp);
+        return -1;
+    }
+
+    int formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        fclose(fp);
+        return (-1);
+    }
+
+    if (_gridSize.size() != 3) {
+        SetErrMsg("Invalid grid size (must be 3D)");
+        fclose(fp);
+        return (-1);
+    }
```

Comment:
Given the type of variable `_gridSize`, is it possible to not be 3? (Will this condition ever come true?)

Reply:
An oversight.  Fixed.